File: DSA/LeetCode/Hashmaps/350. Intersection of Two Arrays II.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {

    const map1 = new Map();
    const map2 = new Map();
    for (let a of nums1) {
        map1.set(a, 1 + (map1.get(a) ?? 0))
    }

    for (let a of nums2) {
        map2.set(a, 1 + (map2.get(a) ?? 0))
    }

    const res = [];
    for (let [k, v] of map1) {
        let min = Math.min(
            map1.get(k) ?? 0,
            map2.get(k) ?? 0,
        )
        res.push(...Array(min).fill(k))
    }
    return res

};

File: DSA/LeetCode/Hashmaps/350. Intersection of Two Arrays II.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:

        map1 = defaultdict(int)
        map2 = defaultdict(int)
        for a in nums1:
            map1[a] +=1
        

        for a in nums2:
            map2[a] +=1
        

        res = []
        for k, v in map1.items():
            min_val = min(map1.get(k, 0), map2.get(k, 0))
            res.extend([k] * min_val)

        return res
        